### PR TITLE
Support annotated-exception

### PR DESCRIPTION
**Pull request by chris-martin**
Obviates https://github.com/freckle/megarepo/pull/31730 and some recent changes to freckle-app. When I roll back what I did to freckle-app I'll probably want to put an explicit lower bound on this version of `bugsnag`.

Also added some build infrastructure - a newer `stack.yaml`, and some nix stuff that helps me.

### CODE REVIEW

File: stack.yaml [hunk 1]
```diff
@@ -17,6 +17,7 @@ ghc-options:
     -Wno-all-missed-specialisations
     -Wno-missing-exported-signatures
     -Wno-missing-import-lists
+    -Wno-missing-kind-signatures
```
Comment (chris-martin) [inline stack.yaml:20]: Newer GHC added this warning I doubt we want.

**Review by pbrisbin — APPROVED** [review]
Something like `annotatedExceptionBeforeNotify` that sets class/message/stacktrace is yet to come?

**Comment by chris-martin** [comment]
Maybe I should go ahead and throw it into this PR.

**Comment by chris-martin** [comment]
@pbrisbin I'm thinking we should go ahead and remove `Network.Bugsnag.Exception.Parse` altogether, wdyt?

**Comment by pbrisbin** [comment]
> I'm thinking we should go ahead and remove Network.Bugsnag.Exception.Parse altogether, wdyt? Would love to simplify by not trying to parse formatted traces at all

That's a good question, but I think it gets at your point about annotated-exceptions not being standardized as a callstack solution. Given that it's not, I don't think we can take away that existing support for trying to get whatever callstacks we can for users not using annotated-exception.

But I think we can evangelize, and ensure good support for annotated-exception and hope it sees broader adoption. Then, perhaps, we can make it essentially required for this library to offer good stacktraces, and we could drop that support for other methods.

**Comment by chris-martin** [comment]
Okay, I did rip that out already, but I can add it back as another fallback.

**Comment by chris-martin** [comment]
I think this could be ready to go.

**However** - While we're here, might as well also upstream `MetaData` from freckle-app?

**Comment by pbrisbin** [comment]
> might as well also upstream MetaData from freckle-app?

Yeah, any part of it that's not Freckle-specific sounds like a decent fit.

I wonder, if we did the following all internally and automatically,

- Handle `ex` or `AnnotatedException ex` in `updateFromOriginalException`
- Had a default (global, "alwasy") `BeforeNotify` that handled `AnnotatedException` and updated `errorClass`, `message`, `stacktrace` **and** attached any other Annotations as metadata

Then do we ever need to mention `AnnotatedException` in our API? I.e. would `updateFromOriginalAnnotatedException` even ever be useful?

**Comment by chris-martin** [comment]
> attached any other Annotations as metadata

For annotations other than `CallStack` and `MetaData`, just apply `show`, and throw the list into an "annotations" bugsnag tab?

**Comment by chris-martin** [comment]
The only reason to include annotated exceptions into the API is if the user has their own `Annotation` subtypes that they want to be able to report to bugsnag in some a special way.

**Comment by chris-martin** [comment]
Would be cool if https://github.com/freckle/blammo/issues/30 were done :)

**Comment by chris-martin** [comment]
Going to also fourmolu this whole project because I don't know how to run brittany and it's falling over on the CPP

**Comment by pbrisbin** [comment]
I guess I defer to you. Would that be useful?

> For annotations other than CallStack and MetaData, just apply show, and throw the list into an "annotations" bugsnag tab?

> the user has their own Annotation subtypes that they want to be able to report to bugsnag in some a special way. Since that's an "open" type that can hold anything, this seems conceivable

> Would be cool if https://github.com/freckle/blammo/issues/30 were done :)

Sure would.

Makes sense.

**Comment by chris-martin** [comment]
I could imagine if annotated-exception caught on more in the ecosystem at large, you could get libraries relying heavily on annotations rather than the underlying exception to convey information you might want to see in bugsnag. But this is hypothetical. And perhaps that could be remedied outside of this library by catching and re-throwing with those annotations converted to `MetaData`. For now I think just supporting `CallStack` and `MetaData` annotations makes sense until there is a use case to drive the design.

**Review by pbrisbin — APPROVED** [review]

**Pull request merged**